File: src/text-featurizer/src/main/scala/TextFeaturizer.scala
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.

package com.microsoft.ml.spark

import java.util.NoSuchElementException

import org.apache.hadoop.fs.Path
import org.apache.spark.ml.feature._
import org.apache.spark.ml._
import org.apache.spark.ml.param._
import org.apache.spark.ml.util._
import org.apache.spark.sql.{DataFrame, Dataset, Row, SaveMode}
import org.apache.spark.sql.types._
import org.apache.spark.ml.Pipeline
import org.apache.spark.ml.attribute.AttributeGroup

trait TextFeaturizerParams extends MMLParams{

  /** Tokenize the input when set to true
    * @group param
    */
  val useTokenizer = BooleanParam(this, "useTokenizer", "Whether to tokenize the input")

  /** @group getParam */
  final def getUseTokenizer: Boolean = $(useTokenizer)

  /** Indicates whether the regex splits on gaps (true) or matches tokens (false)
    * @group param
    */
  val tokenizerGaps = BooleanParam(
    this,
    "tokenizerGaps",
    "Indicates whether regex splits on gaps (true) or matches tokens (false)."
  )

  /** @group getParam */
  final def getTokenizerGaps: Boolean = $(tokenizerGaps)

  /** Minumum token length; must be 0 or greater.
    * @group param
    */
  val minTokenLength = IntParam(this, "minTokenLength", "Minimum token length, >= 0.")

  /** @group getParam */
  final def getMinTokenLength: Int = $(minTokenLength)

  /** Regex pattern used to match delimiters if gaps (true) or tokens (false)
    * @group param
    */
  val tokenizerPattern = StringParam(
    this,
    "tokenizerPattern",
    "Regex pattern used to match delimiters if gaps is true or tokens if gaps is false.")

  /** @group getParam */
  final def getTokenizerPattern: String = $(tokenizerPattern)

  /** Indicates whether to convert all characters to lowercase before tokenizing.
    * @group param
    */
  val toLowercase = BooleanParam(
    this,
    "toLowercase",
    "Indicates whether to convert all characters to lowercase before tokenizing.")

  /** @group getParam */
  final def getToLowercase: Boolean = $(toLowercase)

  /** Indicates whether to remove stop words from tokenized data.
    * @group param
    */
  val useStopWordsRemover = BooleanParam(this,
    "useStopWordsRemover",
    "Whether to remove stop words from tokenized data")

  /** @group getParam */
  final def getUseStopWordsRemover: Boolean = $(useStopWordsRemover)

  /** Indicates whether a case sensitive comparison is performed on stop words.
    * @group param
    */
  val caseSensitiveStopWords = BooleanParam(
    this,
    "caseSensitiveStopWords",
    " Whether to do a case sensitive comparison over the stop words")

  /** @group getParam */
  final def getCaseSensitiveStopWords: Boolean = $(caseSensitiveStopWords)

  /** Specify the language to use for stop word removal. The Use the custom setting when using the
    * stopWords input
    * @group param
    */
  val defaultStopWordLanguage = StringParam(this,
    "defaultStopWordLanguage",
    "Which language to use for the stop word remover," +
      " set this to custom to use the stopWords input")

  /** @group getParam */
  final def getDefaultStopWordLanguage: String = $(defaultStopWordLanguage)

  /** The words to be filtered out. This is a comma separated list of words, encoded as a single string.
    * For example, "a, the, and"
    */
  val stopWords = StringParam(this, "stopWords", "The words to be filtered out.")

  /** @group getParam */
  final def getStopWords: String = $(stopWords)

  /** Enumerate N grams when set
    * @group param
    */
  val useNGram = BooleanParam(this, "useNGram", "Whether to enumerate N grams")

  /** @group getParam */
  final def getUseNGram: Boolean = $(useNGram)

  /** The size of the Ngrams
    * @group param
    */
  val nGramLength = IntParam(this, "nGramLength", "The size of the Ngrams")

  /** @group getParam */
  final def getNGramLength: Int = $(nGramLength)

  /** All nonnegative word counts are set to 1 when set to true
    * @group param
    */
  val binary = BooleanParam(
    this,
    "binary",
    "If true, all nonegative word counts are set to 1")

  /** @group getParam */
  final def getBinary: Boolean = $(binary)

  /** Set the number of features to hash each document to
    * @group param
    */
  val numFeatures = IntParam(
    this,
    "numFeatures",
    "Set the number of features to hash each document to")

  /** @group getParam */
  final def getNumFeatures: Int = $(numFeatures)

  /** Scale the Term Frequencies by IDF when set to true
    * @group param
    */
  val useIDF = BooleanParam(
    this,
    "useIDF",
    "Whether to scale the Term Frequencies by IDF")

  /** @group getParam */
  final def getUseIDF: Boolean = $(useIDF)

  /** Minimum number of documents in which a term should appear.
    * @group param
    */
  val minDocFreq = IntParam(
    this,
    "minDocFreq",
    "The minimum number of documents in which a term should appear.")

  /** @group getParam */
  final def getMinDocFreq: Int = $(minDocFreq)

}

object TextFeaturizer extends DefaultParamsReadable[TextFeaturizer]

/** Featurize text.
  *
  * The default output column name is "<uid>__output"
  * @param uid The id of the module
  */
class TextFeaturizer(override val uid: String)
  extends Estimator[TextFeaturizerModel]
    with TextFeaturizerParams with HasInputCol with HasOutputCol {
  def this() = this(Identifiable.randomUID("TextFeaturizer"))

  setDefault(outputCol, uid + "__output")

  def setUseTokenizer(value: Boolean): this.type = set(useTokenizer, value)

  setDefault(useTokenizer -> true)

  /** @group setParam */
  def setTokenizerGaps(value: Boolean): this.type = set(tokenizerGaps, value)

  setDefault(tokenizerGaps -> true)

  /** @group setParam */
  def setMinTokenLength(value: Int): this.type = set(minTokenLength, value)

  setDefault(minTokenLength -> 0)

  /** @group setParam */
  def setTokenizerPattern(value: String): this.type =
  set(tokenizerPattern, value)

  setDefault(tokenizerPattern -> "\\s+")

  /** @group setParam */
  def setToLowercase(value: Boolean): this.type = set(toLowercase, value)

  setDefault(toLowercase -> true)

  /** @group setParam */
  def setUseStopWordsRemover(value: Boolean): this.type =
  set(useStopWordsRemover, value)

  setDefault(useStopWordsRemover -> false)

  /** @group setParam */
  def setCaseSensitiveStopWords(value: Boolean): this.type =
  set(caseSensitiveStopWords, value)

  setDefault(caseSensitiveStopWords -> false)

  /** @group setParam */
  def setDefaultStopWordLanguage(value: String): this.type =
  set(defaultStopWordLanguage, value)

  setDefault(defaultStopWordLanguage -> "english")

  /** @group setParam */
  def setStopWords(value: String): this.type = set(stopWords, value)

  /** @group setParam */
  def setUseNGram(value: Boolean): this.type = set(useNGram, value)

  /** @group setParam */
  def setNGramLength(value: Int): this.type = set(nGramLength, value)

  setDefault(useNGram -> false, nGramLength -> 2)

  /** @group setParam */
  def setBinary(value: Boolean): this.type = set(binary, value)

  /** @group setParam */
  def setNumFeatures(value: Int): this.type = set(numFeatures, value)

  setDefault(numFeatures -> (1 << 18), binary -> false)

  /** @group setParam */
  def setUseIDF(value: Boolean): this.type = set(useIDF, value)

  /** @group setParam */
  def setMinDocFreq(value: Int): this.type = set(minDocFreq, value)

  setDefault(useIDF -> true, minDocFreq -> 1)

  private def setParamInternal[M <: PipelineStage, T](model: M,
                                                      name: String,
                                                      value: T) = {
    model.set(model.getParam(name), value)
  }

  private def getParamInternal[M <: PipelineStage, T](model: M, name: String) = {
    model.getOrDefault(model.getParam(name))
  }

  override def fit(dataset: Dataset[_]): TextFeaturizerModel = {
    try {
      getUseTokenizer
    } catch {
      case e: NoSuchElementException => setUseTokenizer(needsTokenizer(dataset.schema))
    }

    transformSchema(dataset.schema)
    var models: List[PipelineStage] = Nil
    if (getUseTokenizer)
      models ::= new RegexTokenizer()
        .setGaps(getTokenizerGaps)
        .setPattern(getTokenizerPattern)
        .setMinTokenLength(getMinTokenLength)
        .setToLowercase(getToLowercase)
    if (getUseStopWordsRemover) {
      val swr =
        new StopWordsRemover().setCaseSensitive(getCaseSensitiveStopWords)
      if (getDefaultStopWordLanguage == "custom") {
        models ::= swr.setStopWords(getStopWords.split(","))
      } else {
        models ::= swr.setStopWords(
          StopWordsRemover.loadDefaultStopWords(getDefaultStopWordLanguage))
      }
    }
    if (getUseNGram)
      models ::= new NGram().setN(getNGramLength)
    models ::= new HashingTF()
      .setBinary(getBinary)
      .setNumFeatures(getNumFeatures)
    if (getUseIDF)
      models ::= new IDF().setMinDocFreq(getMinDocFreq)
    models = models.reverse

    val chainedModels = models
      .zip(0 to models.length)
      .map(
        { pair: (PipelineStage, Int) =>
          val model = pair._1
          val i = pair._2
          if (i == 0) {
            setParamInternal(model, "inputCol", getInputCol)
          } else if (i < models.length - 1) {
            setParamInternal(model,
              "inputCol",
              getParamInternal(models(i - 1), "outputCol"))
          } else {
            val m1 =
              setParamInternal(model,
                "inputCol",
                getParamInternal(models(i - 1), "outputCol"))
            setParamInternal(m1, "outputCol", getOutputCol)
          }
        }
      )
    val colsToDrop = chainedModels.reverse.tail
      .map(getParamInternal(_, "outputCol").asInstanceOf[String])
    val fitPipeline =
      new Pipeline().setStages(chainedModels.toArray).fit(dataset)
    new TextFeaturizerModel(uid, fitPipeline, colsToDrop).setParent(this)
  }

  override def copy(extra: ParamMap): Estimator[TextFeaturizerModel] =
    defaultCopy(extra)

  def transformSchema(schema: StructType): StructType = {
    val inputType = schema($(inputCol)).dataType
    validateInputType(inputType)
    if (schema.fieldNames.contains($(outputCol))) {
      throw new IllegalArgumentException(
        s"Output column ${$(outputCol)} already exists.")
    }
    val attrGroup = new AttributeGroup($(outputCol), $(numFeatures))
    appendColumn(schema, attrGroup.toStructField())
  }

  private def needsTokenizer(schema: StructType) = {
    val inputType = schema($(inputCol)).dataType
    inputType == StringType
  }

  private def validateInputType(inputType: DataType) = {
    if (getUseTokenizer) {
      if (inputType == ArrayType(StringType)) {
        require(
          inputType == StringType,
          s"Input type must be string type but got $inputType. " +
            s"It looks like your data is already tokenized, Try with useTokenizer=False")
      }
      require(inputType == StringType,
        s"Input type must be string type but got $inputType.")
    } else if (getUseNGram) {
      if (inputType == StringType) {
        require(
          inputType == ArrayType(StringType),
          s"Input type must be Array[string] type but got $inputType. " +
            s"It looks like your data not tokenized, Try with useTokenizer=True")
      }
      require(
        inputType == ArrayType(StringType),
        s"Input type must be Array(String) type type but got $inputType.")
    } else {
      if (inputType == StringType) {
        require(
          inputType.isInstanceOf[ArrayType],
          s"Input type must be Array[_] type but got $inputType. " +
            s"It looks like your data not tokenized, Try with useTokenizer=True")
      }
      require(inputType.isInstanceOf[ArrayType],
        s"Input type must be Array(_) type type but got $inputType.")
    }
  }

  private def appendColumn(schema: StructType, col: StructField): StructType = {
    require(!schema.fieldNames.contains(col.name),
      s"Column ${col.name} already exists.")
    StructType(schema.fields :+ col)
  }
}

class TextFeaturizerModel(val uid: String,
                          fitPipeline: PipelineModel,
                          colsToDrop: List[String])
  extends Model[TextFeaturizerModel] with MLWritable {

  override def write: MLWriter = new TextFeaturizerModel.TextFeaturizerModelWriter(uid, fitPipeline, colsToDrop)

  override def copy(extra: ParamMap): TextFeaturizerModel = defaultCopy(extra)

  override def transform(dataset: Dataset[_]): DataFrame = {
    colsToDrop.foldRight(fitPipeline.transform(dataset))((col, df) =>
      df.drop(col))
  }

  override def transformSchema(schema: StructType): StructType =
    colsToDrop.foldRight(fitPipeline.transformSchema(schema))((col, schema) =>
      StructType(schema.drop(schema.fieldIndex(col))))
}

object TextFeaturizerModel extends MLReadable[TextFeaturizerModel] {

  private val fitPipelinePart = "fitPipeline"
  private val colsToDropPart = "colsToDrop"
  private val dataPart = "data"

  override def read: MLReader[TextFeaturizerModel] = new TextFeaturizerModelReader

  override def load(path: String): TextFeaturizerModel = super.load(path)

  /** [[MLWriter]] instance for [[TextFeaturizerModel]] */
  private[TextFeaturizerModel]
  class TextFeaturizerModelWriter(val uid: String,
                                  val fitPipeline: PipelineModel,
                                  val colsToDrop: List[String])
    extends MLWriter {

    private case class Data(uid: String)

    override protected def saveImpl(path: String): Unit = {
      val overwrite = this.shouldOverwrite
      val qualPath = PipelineUtilities.makeQualifiedPath(sc, path)
      // Required in order to allow this to be part of an ML pipeline
      PipelineUtilities.saveMetadata(uid,
        TextFeaturizerModel.getClass.getName.replace("$", ""),
        new Path(path, "metadata").toString,
        sc,
        overwrite)

      val dataPath = new Path(qualPath, dataPart).toString

      // Save data
      val data = Data(uid)

      // save the columns to drop
      ObjectUtilities.writeObject(colsToDrop, qualPath, colsToDropPart, sc, overwrite)

      // save the pipeline
      val fitPipelinePath = new Path(qualPath, fitPipelinePart).toString
      val fitPipelineWriter =
        if (overwrite) fitPipeline.write.overwrite()
        else fitPipeline.write
      fitPipelineWriter.save(fitPipelinePath)

      val saveMode =
        if (overwrite) SaveMode.Overwrite
        else SaveMode.ErrorIfExists

      sparkSession.createDataFrame(Seq(data)).repartition(1).write.mode(saveMode).parquet(dataPath)
    }
  }

  private class TextFeaturizerModelReader
    extends MLReader[TextFeaturizerModel] {
    override def load(path: String): TextFeaturizerModel = {
      val qualPath = PipelineUtilities.makeQualifiedPath(sc, path)
      // load the uid and one hot encoding param
      val dataPath = new Path(qualPath, dataPart).toString
      val data = sparkSession.read.format("parquet").load(dataPath)
      val Row(uid: String) = data.select("uid").head()

      // load the fit pipeline
      val fitPipelinePath = new Path(qualPath, fitPipelinePart).toString
      val fitPipeline = PipelineModel.load(fitPipelinePath)

      // load the columns to drop
      val colsToDrop = ObjectUtilities.loadObject[List[String]](qualPath, colsToDropPart, sc)

      new TextFeaturizerModel(uid, fitPipeline, colsToDrop)
    }
  }

}

File: src/src/main/resources/mmlspark/TextFeaturizer.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from pyspark import keyword_only
from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.wrapper import JavaTransformer, JavaEstimator, JavaModel
from pyspark.ml.common import inherit_doc
from mmlspark.Utils import *

@inherit_doc
class TextFeaturizer(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaEstimator):
    """
    Featurize text.

    Args:

        binary (bool): If true, all nonegative word counts are set to 1 (default: false)
        caseSensitiveStopWords (bool): Whether to do a case sensitive comparison over the stop words (default: false)
        defaultStopWordLanguage (str): Which language to use for the stop word remover, set this to custom to use the stopWords input (default: english)
        inputCol (str): The name of the input column
        minDocFreq (int): The minimum number of documents in which a term should appear. (default: 1)
        minTokenLength (int): Minimum token length, >= 0. (default: 0)
        nGramLength (int): The size of the Ngrams (default: 2)
        numFeatures (int): Set the number of features to hash each document to (default: 262144)
        outputCol (str): The name of the output column (default: [self.uid]_output)
        stopWords (str): The words to be filtered out.
        toLowercase (bool): Indicates whether to convert all characters to lowercase before tokenizing. (default: true)
        tokenizerGaps (bool): Indicates whether regex splits on gaps (true) or matches tokens (false). (default: true)
        tokenizerPattern (str): Regex pattern used to match delimiters if gaps is true or tokens if gaps is false. (default: \s+)
        useIDF (bool): Whether to scale the Term Frequencies by IDF (default: true)
        useNGram (bool): Whether to enumerate N grams (default: false)
        useStopWordsRemover (bool): Whether to remove stop words from tokenized data (default: false)
        useTokenizer (bool): Whether to tokenize the input (default: true)
    """

    @keyword_only
    def __init__(self, binary=False, caseSensitiveStopWords=False, defaultStopWordLanguage="english", inputCol=None, minDocFreq=1, minTokenLength=0, nGramLength=2, numFeatures=262144, outputCol=None, stopWords=None, toLowercase=True, tokenizerGaps=True, tokenizerPattern="\s+", useIDF=True, useNGram=False, useStopWordsRemover=False, useTokenizer=True):
        super(TextFeaturizer, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.TextFeaturizer")
        self.binary = Param(self, "binary", "binary: If true, all nonegative word counts are set to 1 (default: false)")
        self._setDefault(binary=False)
        self.caseSensitiveStopWords = Param(self, "caseSensitiveStopWords", "caseSensitiveStopWords:  Whether to do a case sensitive comparison over the stop words (default: false)")
        self._setDefault(caseSensitiveStopWords=False)
        self.defaultStopWordLanguage = Param(self, "defaultStopWordLanguage", "defaultStopWordLanguage: Which language to use for the stop word remover, set this to custom to use the stopWords input (default: english)")
        self._setDefault(defaultStopWordLanguage="english")
        self.inputCol = Param(self, "inputCol", "inputCol: The name of the input column")
        self.minDocFreq = Param(self, "minDocFreq", "minDocFreq: The minimum number of documents in which a term should appear. (default: 1)")
        self._setDefault(minDocFreq=1)
        self.minTokenLength = Param(self, "minTokenLength", "minTokenLength: Minimum token length, >= 0. (default: 0)")
        self._setDefault(minTokenLength=0)
        self.nGramLength = Param(self, "nGramLength", "nGramLength: The size of the Ngrams (default: 2)")
        self._setDefault(nGramLength=2)
        self.numFeatures = Param(self, "numFeatures", "numFeatures: Set the number of features to hash each document to (default: 262144)")
        self._setDefault(numFeatures=262144)
        self.outputCol = Param(self, "outputCol", "outputCol: The name of the output column (default: [self.uid]_output)")
        self._setDefault(outputCol=self.uid + "_output")
        self.stopWords = Param(self, "stopWords", "stopWords: The words to be filtered out.")
        self.toLowercase = Param(self, "toLowercase", "toLowercase: Indicates whether to convert all characters to lowercase before tokenizing. (default: true)")
        self._setDefault(toLowercase=True)
        self.tokenizerGaps = Param(self, "tokenizerGaps", "tokenizerGaps: Indicates whether regex splits on gaps (true) or matches tokens (false). (default: true)")
        self._setDefault(tokenizerGaps=True)
        self.tokenizerPattern = Param(self, "tokenizerPattern", "tokenizerPattern: Regex pattern used to match delimiters if gaps is true or tokens if gaps is false. (default: \s+)")
        self._setDefault(tokenizerPattern="\s+")
        self.useIDF = Param(self, "useIDF", "useIDF: Whether to scale the Term Frequencies by IDF (default: true)")
        self._setDefault(useIDF=True)
        self.useNGram = Param(self, "useNGram", "useNGram: Whether to enumerate N grams (default: false)")
        self._setDefault(useNGram=False)
        self.useStopWordsRemover = Param(self, "useStopWordsRemover", "useStopWordsRemover: Whether to remove stop words from tokenized data (default: false)")
        self._setDefault(useStopWordsRemover=False)
        self.useTokenizer = Param(self, "useTokenizer", "useTokenizer: Whether to tokenize the input (default: true)")
        self._setDefault(useTokenizer=True)
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)

    @keyword_only
    def setParams(self, binary=False, caseSensitiveStopWords=False, defaultStopWordLanguage="english", inputCol=None, minDocFreq=1, minTokenLength=0, nGramLength=2, numFeatures=262144, outputCol=None, stopWords=None, toLowercase=True, tokenizerGaps=True, tokenizerPattern="\s+", useIDF=True, useNGram=False, useStopWordsRemover=False, useTokenizer=True):
        """
        Set the (keyword only) parameters

        Args:

            binary (bool): If true, all nonegative word counts are set to 1 (default: false)
            caseSensitiveStopWords (bool): Whether to do a case sensitive comparison over the stop words (default: false)
            defaultStopWordLanguage (str): Which language to use for the stop word remover, set this to custom to use the stopWords input (default: english)
            inputCol (str): The name of the input column
            minDocFreq (int): The minimum number of documents in which a term should appear. (default: 1)
            minTokenLength (int): Minimum token length, >= 0. (default: 0)
            nGramLength (int): The size of the Ngrams (default: 2)
            numFeatures (int): Set the number of features to hash each document to (default: 262144)
            outputCol (str): The name of the output column (default: [self.uid]_output)
            stopWords (str): The words to be filtered out.
            toLowercase (bool): Indicates whether to convert all characters to lowercase before tokenizing. (default: true)
            tokenizerGaps (bool): Indicates whether regex splits on gaps (true) or matches tokens (false). (default: true)
            tokenizerPattern (str): Regex pattern used to match delimiters if gaps is true or tokens if gaps is false. (default: \s+)
            useIDF (bool): Whether to scale the Term Frequencies by IDF (default: true)
            useNGram (bool): Whether to enumerate N grams (default: false)
            useStopWordsRemover (bool): Whether to remove stop words from tokenized data (default: false)
            useTokenizer (bool): Whether to tokenize the input (default: true)
        """
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        return self._set(**kwargs)

    def setBinary(self, value):
        """

        Args:

            binary (bool): If true, all nonegative word counts are set to 1 (default: false)

        """
        self._set(binary=value)
        return self


    def getBinary(self):
        """

        Returns:

            bool: If true, all nonegative word counts are set to 1 (default: false)
        """
        return self.getOrDefault(self.binary)


    def setCaseSensitiveStopWords(self, value):
        """

        Args:

            caseSensitiveStopWords (bool): Whether to do a case sensitive comparison over the stop words (default: false)

        """
        self._set(caseSensitiveStopWords=value)
        return self


    def getCaseSensitiveStopWords(self):
        """

        Returns:

            bool: Whether to do a case sensitive comparison over the stop words (default: false)
        """
        return self.getOrDefault(self.caseSensitiveStopWords)


    def setDefaultStopWordLanguage(self, value):
        """

        Args:

            defaultStopWordLanguage (str): Which language to use for the stop word remover, set this to custom to use the stopWords input (default: english)

        """
        self._set(defaultStopWordLanguage=value)
        return self


    def getDefaultStopWordLanguage(self):
        """

        Returns:

            str: Which language to use for the stop word remover, set this to custom to use the stopWords input (default: english)
        """
        return self.getOrDefault(self.defaultStopWordLanguage)


    def setInputCol(self, value):
        """

        Args:

            inputCol (str): The name of the input column

        """
        self._set(inputCol=value)
        return self


    def getInputCol(self):
        """

        Returns:

            str: The name of the input column
        """
        return self.getOrDefault(self.inputCol)


    def setMinDocFreq(self, value):
        """

        Args:

            minDocFreq (int): The minimum number of documents in which a term should appear. (default: 1)

        """
        self._set(minDocFreq=value)
        return self


    def getMinDocFreq(self):
        """

        Returns:

            int: The minimum number of documents in which a term should appear. (default: 1)
        """
        return self.getOrDefault(self.minDocFreq)


    def setMinTokenLength(self, value):
        """

        Args:

            minTokenLength (int): Minimum token length, >= 0. (default: 0)

        """
        self._set(minTokenLength=value)
        return self


    def getMinTokenLength(self):
        """

        Returns:

            int: Minimum token length, >= 0. (default: 0)
        """
        return self.getOrDefault(self.minTokenLength)


    def setNGramLength(self, value):
        """

        Args:

            nGramLength (int): The size of the Ngrams (default: 2)

        """
        self._set(nGramLength=value)
        return self


    def getNGramLength(self):
        """

        Returns:

            int: The size of the Ngrams (default: 2)
        """
        return self.getOrDefault(self.nGramLength)


    def setNumFeatures(self, value):
        """

        Args:

            numFeatures (int): Set the number of features to hash each document to (default: 262144)

        """
        self._set(numFeatures=value)
        return self


    def getNumFeatures(self):
        """

        Returns:

            int: Set the number of features to hash each document to (default: 262144)
        """
        return self.getOrDefault(self.numFeatures)


    def setOutputCol(self, value):
        """

        Args:

            outputCol (str): The name of the output column (default: [self.uid]_output)

        """
        self._set(outputCol=value)
        return self


    def getOutputCol(self):
        """

        Returns:

            str: The name of the output column (default: [self.uid]_output)
        """
        return self.getOrDefault(self.outputCol)


    def setStopWords(self, value):
        """

        Args:

            stopWords (str): The words to be filtered out.

        """
        self._set(stopWords=value)
        return self


    def getStopWords(self):
        """

        Returns:

            str: The words to be filtered out.
        """
        return self.getOrDefault(self.stopWords)


    def setToLowercase(self, value):
        """

        Args:

            toLowercase (bool): Indicates whether to convert all characters to lowercase before tokenizing. (default: true)

        """
        self._set(toLowercase=value)
        return self


    def getToLowercase(self):
        """

        Returns:

            bool: Indicates whether to convert all characters to lowercase before tokenizing. (default: true)
        """
        return self.getOrDefault(self.toLowercase)


    def setTokenizerGaps(self, value):
        """

        Args:

            tokenizerGaps (bool): Indicates whether regex splits on gaps (true) or matches tokens (false). (default: true)

        """
        self._set(tokenizerGaps=value)
        return self


    def getTokenizerGaps(self):
        """

        Returns:

            bool: Indicates whether regex splits on gaps (true) or matches tokens (false). (default: true)
        """
        return self.getOrDefault(self.tokenizerGaps)


    def setTokenizerPattern(self, value):
        """

        Args:

            tokenizerPattern (str): Regex pattern used to match delimiters if gaps is true or tokens if gaps is false. (default: \s+)

        """
        self._set(tokenizerPattern=value)
        return self


    def getTokenizerPattern(self):
        """

        Returns:

            str: Regex pattern used to match delimiters if gaps is true or tokens if gaps is false. (default: \s+)
        """
        return self.getOrDefault(self.tokenizerPattern)


    def setUseIDF(self, value):
        """

        Args:

            useIDF (bool): Whether to scale the Term Frequencies by IDF (default: true)

        """
        self._set(useIDF=value)
        return self


    def getUseIDF(self):
        """

        Returns:

            bool: Whether to scale the Term Frequencies by IDF (default: true)
        """
        return self.getOrDefault(self.useIDF)


    def setUseNGram(self, value):
        """

        Args:

            useNGram (bool): Whether to enumerate N grams (default: false)

        """
        self._set(useNGram=value)
        return self


    def getUseNGram(self):
        """

        Returns:

            bool: Whether to enumerate N grams (default: false)
        """
        return self.getOrDefault(self.useNGram)


    def setUseStopWordsRemover(self, value):
        """

        Args:

            useStopWordsRemover (bool): Whether to remove stop words from tokenized data (default: false)

        """
        self._set(useStopWordsRemover=value)
        return self


    def getUseStopWordsRemover(self):
        """

        Returns:

            bool: Whether to remove stop words from tokenized data (default: false)
        """
        return self.getOrDefault(self.useStopWordsRemover)


    def setUseTokenizer(self, value):
        """

        Args:

            useTokenizer (bool): Whether to tokenize the input (default: true)

        """
        self._set(useTokenizer=value)
        return self


    def getUseTokenizer(self):
        """

        Returns:

            bool: Whether to tokenize the input (default: true)
        """
        return self.getOrDefault(self.useTokenizer)



    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.TextFeaturizer"

    @staticmethod
    def _from_java(java_stage):
        module_name=TextFeaturizer.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".TextFeaturizer"
        return from_java(java_stage, module_name)

    def _create_model(self, java_model):
        return TextFeaturizerModel(java_model)


class TextFeaturizerModel(ComplexParamsMixin, JavaModel, JavaMLWritable, JavaMLReadable):
    """
    Model fitted by :class:`TextFeaturizer`.

    This class is left empty on purpose.
    All necessary methods are exposed through inheritance.
    """

    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.TextFeaturizerModel"

    @staticmethod
    def _from_java(java_stage):
        module_name=TextFeaturizerModel.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".TextFeaturizerModel"
        return from_java(java_stage, module_name)

